pkg/controllers/user/pipeline/controller/pipeline/pipeline.go

```diff
@@ -0,0 +1,155 @@
+package pipeline
+
+import (
+	"context"
+	"errors"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/remote"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/satori/uuid"
+	"github.com/sirupsen/logrus"
+)
+
+type Lifecycle struct {
+	pipelines                  v3.PipelineInterface
+	pipelineLister             v3.PipelineLister
+	sourceCodeCredentialLister v3.SourceCodeCredentialLister
+}
+
+func Register(ctx context.Context, cluster *config.UserContext) {
+	pipelines := cluster.Management.Management.Pipelines("")
+	pipelineLister := pipelines.Controller().Lister()
+	pipelineExecutions := cluster.Management.Management.PipelineExecutions("")
+	pipelineExecutionLogs := cluster.Management.Management.PipelineExecutionLogs("")
+	sourceCodeCredentialLister := cluster.Management.Management.SourceCodeCredentials("").Controller().Lister()
+
+	pipelineLifecycle := &Lifecycle{
+		pipelines:                  pipelines,
+		pipelineLister:             pipelineLister,
+		sourceCodeCredentialLister: sourceCodeCredentialLister,
+	}
+	s := &CronSyncer{
+		pipelineLister:        pipelineLister,
+		pipelines:             pipelines,
+		pipelineExecutions:    pipelineExecutions,
+		pipelienExecutionLogs: pipelineExecutionLogs,
+	}
+
+	pipelines.AddLifecycle("pipeline-controller", pipelineLifecycle)
+	go s.sync(ctx, syncInterval)
+}
+
+func (l *Lifecycle) Create(obj *v3.Pipeline) (*v3.Pipeline, error) {
+
+	obj.Status.NextRun = 1
+	obj.Status.State = "active"
+	if obj.Status.Token == "" {
+		//random token for webhook validation
+		obj.Status.Token = uuid.NewV4().String()
+	}
+
+	if obj.Spec.TriggerWebhook && obj.Status.WebHookID == "" {
+		id, err := l.createHook(obj)
+		if err != nil {
+			return obj, err
+		}
+		obj.Status.WebHookID = id
+		if _, err := l.pipelines.Update(obj); err != nil {
+			return obj, err
+		}
+	}
+	return obj, nil
+}
+
+func (l *Lifecycle) Updated(obj *v3.Pipeline) (*v3.Pipeline, error) {
+	previous, err := l.pipelineLister.Get(obj.Namespace, obj.Name)
+	if err != nil {
+		return obj, err
+	}
+	//handle cron update
+	if (obj.Spec.TriggerCronExpression != previous.Spec.TriggerCronExpression) ||
+		(obj.Spec.TriggerCronTimezone != previous.Spec.TriggerCronTimezone) {
+		//cron trigger changed, reset
+		obj.Status.NextStart = ""
+	}
+
+	//handle webhook
+	if previous.Spec.TriggerWebhook && previous.Status.WebHookID != "" && !obj.Spec.TriggerWebhook {
+		if err := l.deleteHook(previous); err != nil {
+			logrus.Errorf("fail to delete previous set webhook")
+		}
+	} else if !previous.Spec.TriggerWebhook && obj.Spec.TriggerWebhook && obj.Status.WebHookID == "" {
+		id, err := l.createHook(obj)
+		if err != nil {
+			return obj, err
+		}
+		obj.Status.WebHookID = id
+		if _, err := l.pipelines.Update(obj); err != nil {
```

Comment:
I am not sure if we should call Update here.  I tried to use Create func and it will do the update automatically, but I haven't try Update. Maybe you can take a try.